common/transport/coap/src/main/java/org/thingsboard/server/transport/coap/CoapTransportResource.java

```diff
@@ -458,5 +469,31 @@ public void removedObserveRelation(ObserveRelation relation) {
         }
     }
 
+    private TbCoapDtlsDeviceAddr getCoapDtlsDeviceAddr(EndpointContext endpointContext) {
+        InetSocketAddress peerAddress = endpointContext.getPeerAddress();
+        String certPemStr = getCertPem(endpointContext);
+        if (StringUtils.isNotBlank(certPemStr)) {
+            for (var entry : dtlsSessionsMap.entrySet()) {
+                TbCoapDtlsDeviceAddr tbCoapDtlsDeviceAddr = entry.getKey();
+                TbCoapDtlsSessionInfo sessionInfo = entry.getValue();
+                String credentials = sessionInfo.getMsg().getCredentials();
+                if (tbCoapDtlsDeviceAddr.getInetSocketAddress().equals(peerAddress) && certPemStr.equals(credentials)) {
+                    return tbCoapDtlsDeviceAddr;
+                }
+            }
+        }
```

Comment:
I don't like the idea of iterating all sessions on each request.
We could leave the key in the dtlsSessionsMap as InetSocketAddress, but the value would be a list of conflicting sessions:
`ConcurrentMap<InetSocketAddress, List<TbCoapDtlsSessionInfo>> dtlsSessionsMap;`
This way, we would need to only iterate a small list, not the whole map.

Reply:
TbCoapDtlsSessionKey tbCoapDtlsSessionKey = new TbCoapDtlsSessionKey(peerAddress, certPemStr); return dtlsSessionsMap.get(key);